### PR TITLE
Reland "Single tap on the previous selection should toggle the toolbar on iOS #108913"

### LINKED ISSUE #111660: Forced Null Dereference causing crash in editableTextState
Crash seen below. Likely culprit PR here: https://github.com/flutter/flutter/pull/108913.

Stack Trace:
at EditableTextState.toggleToolbar (editable_text.dart:3320 )
at TextSelectionGestureDetectorBuilder.onSingleTapUp (text_selection.dart:1885 )
at _TextFieldSelectionGestureDetectorBuilder.onSingleTapUp (text_field.dart:92 )
at _TextFieldSelectionGestureDetectorBuilder.onSingleTapUp (text_field.dart:90 )
at _TextSelectionGestureDetectorState._handleTapUp (text_selection.dart:2327 )
at _TextSelectionGestureDetectorState._handleTapUp (text_selection.dart:2325 )
at TapGestureRecognizer.handleTapUp.<anonymous closure> (tap.dart:624 )
at GestureRecognizer.invokeCallback (recognizer.dart:253 )
at TapGestureRecognizer.handleTapUp (tap.dart:624 )
at BaseTapGestureRecognizer._checkUp (tap.dart:306 )
at BaseTapGestureRecognizer.acceptGesture (tap.dart:276 )
at GestureArenaManager.sweep (arena.dart:163 )
at GestureBinding.handleEvent (binding.dart:464 )
at GestureBinding.dispatchEvent (binding.dart:440 )
at RendererBinding.dispatchEvent (binding.dart:333 )
at GestureBinding._handlePointerEventImmediately (binding.dart:395 )
at GestureBinding._flushPointerEventQueue (binding.dart:357 )
at GestureBinding._handlePointerDataPacket (binding.dart:295 )
at GestureBinding._handlePointerDataPacket (binding.dart:290 )
at _invoke1 (hooks.dart:164 )
at PlatformDispatcher._dispatchPointerDataPacket (platform_dispatcher.dart:339 )
at _dispatchPointerDataPacket (hooks.dart:91 )

### PR DESCRIPTION
**Pull request by Renzo-Olivares**
Relands #108913

Fixes #111660

Was failing due to a call to EditableText.toggleToolbar which does a forced null dereference on SelectionOverlay. There is an assert that checks if the SelectionOverlay exists, but this does not work on production. This change creates a SelectionOverlay if it does not exist when toggleToolbar is called and adds a regression test.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -3315,10 +3315,13 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
   }
 
   /// Toggles the visibility of the toolbar.
-  void toggleToolbar() {
-    assert(_selectionOverlay != null);
+  void toggleToolbar([bool hideHandles = true]) {
+    if (_selectionOverlay == null) {
+      _createSelectionOverlay();
+    }
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:3321]: `final selectionOverlay = _selectionOverlay ??= createSelectionOverlay()` to avoid the force cast?

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 2]
```diff
@@ -1904,7 +1919,16 @@ class TextSelectionGestureDetectorBuilder {
...
             case PointerDeviceKind.unknown:
               // On iOS/iPadOS a touch tap places the cursor at the edge of the word.
-              renderEditable.selectWordEdge(cause: SelectionChangedCause.tap);
+              final TextSelection previousSelection = editableText.textEditingValue.selection;
+              // If the tap was within the previous selection, then the selection should stay the same.
+              if (!_tapWasOnSelection(details.globalPosition)) {
+                renderEditable.selectWordEdge(cause: SelectionChangedCause.tap);
+              }
+              if (previousSelection == editableText.textEditingValue.selection && renderEditable.hasFocus) {
+                editableText.toggleToolbar(false);
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/text_selection.dart:1928]: Could this be combined with the previous if clause?

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 3]
```diff
@@ -1634,6 +1634,19 @@ class TextSelectionGestureDetectorBuilder {
...
+  bool _tapWasOnSelection(Offset position) {
+    if (renderEditable.selection == null) {
+      return false;
+    }
+
+    final TextPosition textPosition = renderEditable.getPositionForPoint(
+      position,
+    );
+
+    return renderEditable.selection!.start < textPosition.offset
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:1646]: Ah I saw the comment from the previous pull request. But you may still want to check the affinity of the incoming `position`?

[the remark this answers is not in the document]
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart]: What should the `TextAffinity` be checked for?

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 2 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:1929]: Not sure if it can, but I'll try it out.

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart]: Oh nvm you don't have to since the logic will be identical.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 4]
```diff
@@ -3316,10 +3318,11 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
   }
 
   /// Toggles the visibility of the toolbar.
-  void toggleToolbar() {
-    assert(_selectionOverlay != null);
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:3320]: I don't know why we didn't create the selection overlay here in the first place.  Seems fine to me.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 5]
```diff
@@ -2718,8 +2718,8 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     _selectionOverlay?.updateForScroll();
   }
 
-  void _createSelectionOverlay() {
-    _selectionOverlay = TextSelectionOverlay(
+  TextSelectionOverlay _createSelectionOverlay() {
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:2721]: I think I like this pattern better 👍

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 6]
```diff
@@ -1634,6 +1634,20 @@ class TextSelectionGestureDetectorBuilder {
...
+    if (selection == null) {
+      return false;
+    }
+
+    final TextPosition textPosition = renderEditable.getPositionForPoint(
+      position,
+    );
+
+    return selection.start < textPosition.offset
+        && selection.end > textPosition.offset;
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/text_selection.dart:1648]: Not inclusive?

**Review by justinmc — APPROVED** [review]
LGTM 👍

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 6 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:1648]: I updated the logic.

File: packages/flutter/test/material/text_field_test.dart [hunk 7]
```diff
@@ -8959,11 +9084,12 @@ void main() {
 
       await tester.tapAt(textfieldStart + const Offset(150.0, 9.0));
       await tester.pump(const Duration(milliseconds: 50));
-      // First tap moved the cursor.
+      // First tap moved the cursor and hides the toolbar.
```
Comment (justinmc) [inline packages/flutter/test/material/text_field_test.dart:9087]: "hides" => "hid"

**Review by justinmc — APPROVED** [review]
Renewing my LGTM after the previous changes :+1:

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 8]
```diff
@@ -1904,7 +1934,29 @@ class TextSelectionGestureDetectorBuilder {
...
+              // editable is focused, then the toolbar is toggled.
+              //
+              // If the `previousSelection` is not collapsed, the tap was on the selection (excluding the bounds, we exclude the
+              // the bounds so we can still tap on the selection edges to select the word edge), and the editable is focused, then
+              // the toolbar is toggled.
+              //
+              // If the tap was neither on the selection exclusively or inclusively then we select the word edge. After
+              // selecting the word edge, if our selection remains the same, then we toggle the toolbar. If the selection changes
+              // then we hide the toolbar.
+              if (((_tapWasOnSelectionExclusive(details.globalPosition) && !previousSelection.isCollapsed)
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/text_selection.dart:1948]: This is a bit hard to read. Could you simplify this a bit?

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:1948]: Also you probably don't need `_tapWasOnSelectionInclusive`? if `previousSelection == editableText.textEditingValue.selection` then we'll toggle the toolbar anyway?

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:1948]: I think there is still the case where the cursor is in the middle of a word `wor|d`, and we tap it.

If we remove the `inclusive` check, then the `exclusive` one will not catch it, and the cursor will move to the word edge.

The expected behavior would be to toggle the toolbar when we tap the cursor.

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:1948]:
```dart
if () {
   // If the tap was on the selection (different meaning for collapsed vs uncollapsed selection) and the editable was focused, then toggle the toolbar.
} else if () {
   // Tap was not on the selection at all.
   // We should still toggle the toolbar, if the selection remains the same after the word edge is selected.
   // If the selection changes then we hide the toolbar.
}
```

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:1946]: I updated the logic a little, getting rid of the `else if`.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:1946]: > Tap was not on the selection at all.
> if the selection remains the same after the word edge is selected.
can the two both be true?

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:1946]: Yeah, this can happen in the following example

'This is an example'

tap the position where 'p' is.
-> Cursor will move to closest word edge `example|`

tap the position where 'p' is again.
-> Cursor will move to closest word edge 'example|'
-> The previousSelection is the same as the new selection after the closest word edge is selected, so the toolbar toggles

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:1946]: For the collapsed tap check you'll have to take text affinity into account I think. For instance if the cursor is at the end of a soft-wrapped line and you tap the start of the next line, you'll get a TextPosition that has the same offset as the caret location.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:1946]: Hmm but if the cursor is located in a place when the affinity doesn't matter then that will be too strict. I guess we can fix that later.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:1946]: Also could you explain why `RenderEditable.hasFocus` need to be taken into account?

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:1946]: Yeah I’ll keep a lookout of affinity issues.

On native if the renderEditable doesn’t have focus we always select the closest word edge. And the toolbar is not toggled unless the editable has focus.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:1946]: Could you check if you tap the latter half of the character before the caret? I think that will give you (x-1, upstream) instead of (x, downstream) so `wasTapOnSelection` will return false? But that should also toggle the toolbar right?

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:1946]: Good point, If i do the later half of the character before the character the selection is upstream. If I then follow that up with a tap on the former half of the character after the caret, then that selection is downstream. Then because of `previousSelection == editableText.textEditingValue.selection`, the toolbar is not toggled in this case.

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:1946]: I think in the case where we select the word edge the affinity does not matter since a word will not be wrapped between two lines (will either fall on first or second line).

In the case where we check for `wasTapOnSelectionInclusive` we do care about the affinity because we want to be able to select the text at beginning of the next line. So we want it to return false when the affinity is different.

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:1949]: There might be a trade off we have to make here since we cannot differentiate why the TextAffinity changed.

We want to toggle the toolbar when
- We call `selectWordEdge` and the selection after the call is the same as the previous selection (including TextAffinity).
- We call `selectWordEdge` and the selection after the call is the same as the previous selection (excluding TextAffinity), this can happen when we tap 

We want to hide the toolbar when 
- We call `selectWordEdge` and the selection after the call is the same as the previous selection (excluding TextAffinity). This can happen when there is a multi-line text field, there is word wrap and we want to tap on the boundary. Let's say the cursor is at the end of the first line. Tapping on the first line boundary should toggle the toolbar, tapping at the beginning of the second line should hide the toolbar.

We want differentiate between whether `TextAffinity` changed due to tapping on a new line but same selection offset, or if it occurred from tapping on the same line with the same selection offset.

**Review by justinmc — APPROVED** [review]
:+1: I'm on board with the compromise made here, it seems like the best option in the circumstances.  But I want to explore this a little farther just to make sure.

You probably already saw this but the logic that gives the affinity based on the side of a character is all the way down [in paragraph_txt.cc](https://github.com/flutter/engine/blob/aa0dc1826fbad055c47e9d0ce63e31c45f966a10/third_party/txt/src/txt/paragraph_txt.cc#L1933-L1938):

```c++
  double glyph_center = (gp->x_pos.start + gp->x_pos.end) / 2;
  if ((direction == TextDirection::ltr && dx < glyph_center) ||
      (direction == TextDirection::rtl && dx >= glyph_center)) {
    return PositionWithAffinity(gp->code_units.start, DOWNSTREAM);
  } else {
    return PositionWithAffinity(gp->code_units.end, UPSTREAM);
  }
```

So that probably shouldn't change.  Instead, the ideal logic is that on tap, we only care about affinity for toggling the toolbar if the tap is at a soft wordwrap.

Can we know about the location of soft wordwraps at this point in the pipeline?  The text has already been laid out since this is a tap callback, and getPositionForPoint (called already on line 1948) requires the text to be laid out too.  So the information we need is there, but if I'm not mistaken, there is no good way to get this from RenderEditable.

### Tl;dr: The alternative to this compromise
You could add a method to RenderEditable that tells you one way or another if a given TextPosition is at a soft wordwrap or not.  I think you can figure this out using LineMetrics.  If the tap is at a soft wordwrap, only toggle the toolbar if the TextAffinity matches.  If it's not at a soft wordwrap, toggle it regardless of the affinity.

Curious about your thoughts on if this is a correct analysis, and if so if it's worth it or not!  But I'm on board with this PR either way.  Maybe this could even be done in a follow-up PR.

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 9]
```diff
@@ -1903,8 +1933,31 @@ class TextSelectionGestureDetectorBuilder {
...
+              //
+              // Toggle the toolbar when the tap is exclusively within the bounds of a non-collapsed `previousSelection`,
+              // and the editable is focused.
+              //
+              // Selects the word edge closest to the tap when the editable is not focused, or if the tap was neither exclusively
+              // or inclusively on `previousSelection`. If the selection remains the same after selecting the word edge, then we
+              // toggle the toolbar. If the selection changes then we hide the toolbar.
+              final TextSelection previousSelection = renderEditable.selection ?? editableText.textEditingValue.selection;
+              final bool isAffinityTheSame = renderEditable.getPositionForPoint(details.globalPosition).affinity == previousSelection.affinity;
+              if (((_tapWasOnSelectionExclusive(details.globalPosition) && !previousSelection.isCollapsed)
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/text_selection.dart:1949]: nit: You're potentially calling `getPositionForPoint` 3 times. The call is via FFI so could be a bit costly than usual dart calls. So maybe store that in a local variable and that might help simplify the code too.

**Review by LongCatIsLooong — APPROVED** [review]

**Comment by Renzo-Olivares** [comment]
> 👍 I'm on board with the compromise made here, it seems like the best option in the circumstances. But I want to explore this a little farther just to make sure.
> 
> You probably already saw this but the logic that gives the affinity based on the side of a character is all the way down [in paragraph_txt.cc](https://github.com/flutter/engine/blob/aa0dc1826fbad055c47e9d0ce63e31c45f966a10/third_party/txt/src/txt/paragraph_txt.cc#L1933-L1938):
> 
> ```c++
>   double glyph_center = (gp->x_pos.start + gp->x_pos.end) / 2;
>   if ((direction == TextDirection::ltr && dx < glyph_center) ||
>       (direction == TextDirection::rtl && dx >= glyph_center)) {
>     return PositionWithAffinity(gp->code_units.start, DOWNSTREAM);
>   } else {
>     return PositionWithAffinity(gp->code_units.end, UPSTREAM);
>   }
> ```
> 
> So that probably shouldn't change. Instead, the ideal logic is that on tap, we only care about affinity for toggling the toolbar if the tap is at a soft wordwrap.
> 
> Can we know about the location of soft wordwraps at this point in the pipeline? The text has already been laid out since this is a tap callback, and getPositionForPoint (called already on line 1948) requires the text to be laid out too. So the information we need is there, but if I'm not mistaken, there is no good way to get this from RenderEditable.
> 
> ### Tl;dr: The alternative to this compromise
> You could add a method to RenderEditable that tells you one way or another if a given TextPosition is at a soft wordwrap or not. I think you can figure this out using LineMetrics. If the tap is at a soft wordwrap, only toggle the toolbar if the TextAffinity matches. If it's not at a soft wordwrap, toggle it regardless of the affinity.
> 
> Curious about your thoughts on if this is a correct analysis, and if so if it's worth it or not! But I'm on board with this PR either way. Maybe this could even be done in a follow-up PR.

I 100% agree with your analysis. I will investigate further on how feasible adding such an API to renderEditable is, but it definitely sounds worth it to add if its possible. That will be in a separate PR. I created an issue for this #113228.
